### ISSUE TITLE
Rule enhancement: "key-spacing" does not consider tab as a whitespace

### ISSUE THREAD

**Issue by doberkofler** [eslint/eslint#1727]
The rule key-spacing is already great but unfortunately it does not correctly work when using tab instead of space as a whitespace.

This causes false positives when
* aligning using tab
* when mixing tabs and spaces

Example (`"key-spacing": [2, {"align": "value", "beforeColon": false, "afterColon": true}]`):
```
var obj1 = {
	prop:<tab>'value',
	p:<tab><tab>'value'
};

3:5  error  Missing space before value for key "p"  key-spacing
```

Example (`"key-spacing": [2, {"beforeColon": false, "afterColon": true}]`):

```
var obj2 = {
	prop:<space><space><space>'value',
	p:<tab><tab>'value'
};

3:4  error  Missing space before value for key "p"  key-spacing
```

To keep the rule as generic as possible, it would be great to have an additional option "tab" ´[2, {"align": "value", "beforeColon": false, "afterColon": true, "tab": 4}]` that can either be undefined to keep the exiting behavior or be set to the number of spaces it should expand to. If it is set it should treat tab like space when not using align alignment and expand to the given number of spaces when using align.

**Comment by nzakas** [comment]
I'm not sure that this can be made to work properly with tabs. Tabs can be the equivalent of one space, our four, depending on what comes before it. Of someone wants to try it, I'll happily review.

**Comment by doberkofler** [comment]
* I'm sure that this is not easy to implement but in my opinion (I'm really keen on using tabs) it would be quite important for eslint to support space and tab interchangeably.
* I imagine that this problem might also arise when supporting rules for the indentation of code.
* I would have hoped that by specifying the tab width, it should be possible to convert the tabs back to spaces but I must admit that I have very little experience in this area.

**Comment by jonathanKingston** [comment]
@doberkofler are you asking for mixed tabs and spaces for alignment?

There is a good reason why this rule exists: https://github.com/eslint/eslint/blob/master/docs/rules/no-mixed-spaces-and-tabs.md

As @nzakas said it depends on the editors width how they will align. So rather than just forcing a coding style on developers, you will be forcing them to set a display width for their tabs too.

---

However I do think pure tabs could be fixed as you suggested by specifying a width, I think that could be in the rule. The change should be pretty trivial to modify with width based upon a parameter by making `getPropertyWhitespace` replace tabs with the specified number of spaces.

**Comment by nzakas** [comment]
The tricky part is that a tab isn't always the same width even on the same line. Just replacing each tab with four spaces doesn't actually tell you whether two things on different lines are aligned because it matters what came before it. Suppose you set tabs to be four spaces. If one property is two characters then the first tab after it is only two spaces. If a property is three characters then the next tab is just one space. So just doing a replacement isn't enough, there's more calculation that has to be used.

**Comment by doberkofler** [comment]
@jonathanKingston No, I'm not asking for mixed tabs and spaces (in one line) for alignment. I would only like to have tabs expanded.
In the following (hopefully better) example, I would like to have the tabs expanded to 4 spaces and therefore not reporting any errors:
```
/*eslint key-spacing: [2, {"align": "value", "beforeColon": false, "afterColon": true, "tab": 4}]*/

var obj1 = {
	prop:<tab>'value',
	p:<tab><tab>'value'
};
var obj2 = {
	prop:<space><space><space>'value',
	p:<tab><tab>'value'
};
```

@nzakas I understand that the tabs cannot simply be replaced by a certain number of spaces. What I wanted to suggest, was to expand the tabs to spaces before the rule checks the alignment.
It did not test it properly but I would imagine some code like this:
```
	function convertTabsToSpaces(line, tabWidth) {
		var chars = line.split(''),
			len = chars.length,
			newLine = '',
			spaces,
			i,
			j;

		for (i = 0; i < len; i++) {
			if (chars[i] === '\t') {
				spaces = 4 - (newLine.length % tabWidth);
				for (j = 0; j < spaces; j++) {
					newLine += ' ';
				}
			} else {
				newLine += chars[i];
			}
		}

		return newLine;
	}
```

**Comment by nzakas** [comment]
Yup, I got that, I was just commenting on it for anyone who might attempt this.

**Comment by doberkofler** [comment]
Excellent, but why is the issue is still marked as question.

**Comment by nzakas** [comment]
Because I dying always have time to adjust labels. Fixed.

**Comment by platinumazure** [comment]
Closing as there has been no movement on this in nearly a year. Feel free to reopen as needed.
